File: content/02-two-pointers/container_with_most_water.py
"""
Problem: Container With Most Water
LeetCode: #11 (Medium) | NeetCode 150 | Blind 75
"""

def max_area(height: list[int]) -> int:
    """
    Time Complexity: O(N)
    Space Complexity: O(1)
    """
    left, right = 0, len(height) - 1
    max_w = 0
    
    while left < right:
        w = right - left
        h = min(height[left], height[right])
        max_w = max(max_w, w * h)
        
        if height[left] < height[right]:
            left += 1
        else:
            right -= 1
            
    return max_w

if __name__ == "__main__":
    assert max_area([1, 8, 6, 2, 5, 4, 8, 3, 7]) == 49
    assert max_area([1, 1]) == 1
    print("Python: max_area passed tests!")

File: content/02-two-pointers/container_with_most_water.js
/**
 * Problem: Container With Most Water
 * LeetCode: #11 (Medium) | NeetCode 150 | Blind 75
 */

function maxArea(height) {
    /**
     * Time Complexity: O(N)
     * Space Complexity: O(1)
     */
    let left = 0, right = height.length - 1;
    let maxW = 0;
    
    while (left < right) {
        const w = right - left;
        const h = Math.min(height[left], height[right]);
        maxW = Math.max(maxW, w * h);
        
        if (height[left] < height[right]) {
            left++;
        } else {
            right--;
        }
    }
    
    return maxW;
}

if (require.main === module) {
    const assert = require('assert');
    assert.strictEqual(maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7]), 49);
    assert.strictEqual(maxArea([1, 1]), 1);
    console.log("JS: maxArea passed tests!");
}
